waltid-services/waltid-e2e-tests/src/test/kotlin/LspPotentialIssuance.kt

```diff
@@ -0,0 +1,384 @@
+      Url(authResp.headers[HttpHeaders.Location]!!)
+
+    } else {
+      // TODO: check PAR support, also check issuer of panasonic (https://mdlpilot.japaneast.cloudapp.azure.com:8017), which uses PAR and doesn't seem to work
+      val parResp = client.submitForm(oauthMetadata.pushedAuthorizationRequestEndpoint!!, parametersOf(
+        authReq.toHttpParameters()
+      ))
+      val authResp = client.get(oauthMetadata.authorizationEndpoint!!) {
+        parameter("request_uri", Json.parseToJsonElement(parResp.bodyAsText()).jsonObject["request_uri"]!!.jsonPrimitive.content)
+      }
+      assertEquals(HttpStatusCode.Found, authResp.status)
+      Url(authResp.headers[HttpHeaders.Location]!!)
+    }
+    assertContains(location.parameters.names(), "code")
+    // ### step 16-18: access token retrieval
+    val tokenResp = client.submitForm(oauthMetadata.tokenEndpoint!!,
+      parametersOf(
+        TokenRequest(
+          GrantType.authorization_code, authReq.clientId,
+          authReq.redirectUri, location.parameters["code"]!!,
+          codeVerifier = codeVerifier
+        ).toHttpParameters())).let { TokenResponse.fromJSON(it.body<JsonObject>()) }
+    assertNotNull(tokenResp.accessToken)
+    assertTrue(tokenResp.isSuccess)
+
+    val deviceKeyId = "device-key"
+    val deviceKeyPair: ECKey = runBlocking { ECKey.parseFromPEMEncodedObjects(FileReader("./src/test/resources/lsp/ec_device_key.pem").readText()).toECKey() }
+    val certificateChain: String = FileReader("./src/test/resources/lsp/cert-device-potential.pem").readText()
+    // ### steps 19-22: credential issuance
+
+    repeat(3) { it ->
+      // TODO: move COSE signing functionality to crypto lib?
+      val coseKey = OneKey(deviceKeyPair.toECPublicKey(), null).AsCBOR().EncodeToBytes()
+      val credReq = CredentialRequest.forOfferedCredential(
+        offeredCredential, ProofOfPossession.CWTProofBuilder(
+          issuerUrl = parsedOffer.credentialIssuer, clientId = authReq.clientId, nonce = tokenResp.cNonce,
+          coseKey = when (it) { // test with COSE_Key header
+            0 -> coseKey
+            else -> null
+          },
+          x5Cert = when (it) { // test with single x509 certificate in x5chain header
+            1 -> certificateChain.encodeToByteArray()
+            else -> null
+          },
+          x5Chain = when (it) { // test with x509 certificate list in x5chain header
+            2 -> listOf(certificateChain.encodeToByteArray())
+            else -> null
+          }
+        ).build(
+          SimpleCOSECryptoProvider(
+            listOf(
+              COSECryptoProviderKeyInfo(
+                deviceKeyId, AlgorithmID.ECDSA_256,
+                deviceKeyPair.toECPublicKey(), deviceKeyPair.toECPrivateKey()
+              )
+            )
+          ), deviceKeyId
+        )
+      )
+
+      val cwt = Cbor.decodeFromByteArray(COSESign1.serializer(), credReq.proof!!.cwt!!.base64UrlDecode())
```

Comment:
ide suggests optin for `@kotlinx.serialization.ExperimentalSerializationApi` is required